### DISCUSSION TITLE
Low resolution after homography transformation

### DISCUSSION THREAD

**Discussion by qiangwang57** [kornia/kornia#1119] [Questions] [answered]
I found that if I use HomographyWarper for the transformation for a large image, the resolution after registration drops significantly, but with cv2, the resolution retains well.

![image](https://user-images.githubusercontent.com/49925149/122764632-94023780-d297-11eb-883a-fa99413e16de.png)

The code is as below:

```
def homographywaper(image, homo_matrix, target_h=256, target_w=256, inverse=False):
    image_tensor = image2tensor(image)
    homo_matrix_tensor = torch.Tensor(homo_matrix)
    homo_matrix_tensor = torch.unsqueeze(homo_matrix_tensor, dim=0)
    if inverse:
        homo_matrix_tensor = torch.inverse(homo_matrix_tensor)
    warper = HomographyWarper(target_h, target_w)
    image_tensor_warped = warper(image_tensor, homo_matrix_tensor)
    image_warped = (tensor2image(image_tensor_warped)*255).astype(np.uint8)
    return image_warped
```

Any tips?

Cheers, Chiang

**Comment by edgarriba** [comment]
@qiangwang57 thanks for reporting this issue. Have you tried with different interpolation modes ? e.g `mode='nearest'`

**Comment by qiangwang57** [comment]
> @qiangwang57 thanks for reporting this issue. Have you tried with different interpolation modes ? e.g `mode='nearest'`

@edgarriba Yes, I tried both, i.e. "nearest" and "biliniear", but no luck.

**Comment by edgarriba** [comment]
What's the original and output image sizes ? /cc @ducha-aiki

**Comment by shijianjian** [comment]
Hi, can you share a minimal Colab to reproduce?

**Comment by qiangwang57** [comment]
> What's the original and output image sizes ? /cc @ducha-aiki

They both are 6000*5800.

**Comment by qiangwang57** [comment]
> Hi, can you share a minimal Colab to reproduce?

@shijianjian 

Here is the code: https://colab.research.google.com/drive/1YNAcYQZ2UWRAuVCZgjj-iHi4dvIyQJgo#scrollTo=gmLGyG0F0aHL
Not sure if viewable as brandnew to colab.

Cheers, Chiang

**Comment by ducha-aiki** [comment]
Thank you! Could you please give me an access? (I have clicked on "request access" button",as it is restricted now)

**Comment by qiangwang57** [comment]
> Thank you! Could you please give me an access? (I have clicked on "request access" button",as it is restricted now)

Sorry that just saw the request. @ducha-aiki @shijianjian you should be able to access it now.

**Comment by ducha-aiki** [comment, marked as answer]
The problem is not in kornia, but in auxilary function, which you defined, which does resize to 256x256:

```python3
def image2tensor(image, size=256):
    if len(image.shape) == 2:
        ## single channel image
        transform = transforms.Compose([transforms.Resize((size, size)),
                                        transforms.ToTensor()])
    else:
        transform = transforms.Compose([transforms.Resize((size, size)),
                                        transforms.ToTensor()])
    pilimage = Image.fromarray(image)
    if pilimage.mode == "RGBA":
        pilimage = pilimage.convert("RGB")

    pilimage = transform(pilimage)
    return pilimage.unsqueeze(0)
```

Instead just use kornia.image_to_tensor in the following way instead of your function:

```python3
def homographywarper(image, homo_matrix, target_h=256, target_w=256, inverse=False):
    image_tensor = kornia.image_to_tensor(image, False).float()/255.
    homo_matrix_tensor = torch.Tensor(homo_matrix)
    homo_matrix_tensor = torch.unsqueeze(homo_matrix_tensor, dim=0)
    if inverse:
        homo_matrix_tensor = torch.inverse(homo_matrix_tensor)
    warper = HomographyWarper(target_h, target_w, mode="nearest", align_corners=True)
    image_tensor_warped = warper(image_tensor, homo_matrix_tensor)
    print(image_tensor_warped.min(), image_tensor_warped.max())
    image_warped = (tensor2image(image_tensor_warped) * 255).astype(np.uint8)
    return image_warped
```

![image](https://user-images.githubusercontent.com/4803565/122968306-b07f9c00-d38b-11eb-9834-e93b1f892743.png)

**Comment by qiangwang57** [comment]
> The problem is not in kornia, but in auxilary function, which you defined, which does resize to 256x256:
> 
> ```python
> def image2tensor(image, size=256):
>     if len(image.shape) == 2:
>         ## single channel image
>         transform = transforms.Compose([transforms.Resize((size, size)),
>                                         transforms.ToTensor()])
>     else:
>         transform = transforms.Compose([transforms.Resize((size, size)),
>                                         transforms.ToTensor()])
>     pilimage = Image.fromarray(image)
>     if pilimage.mode == "RGBA":
>         pilimage = pilimage.convert("RGB")
> 
>     pilimage = transform(pilimage)
>     return pilimage.unsqueeze(0)
> ```
> 
> Instead just use kornia.image_to_tensor in the following way instead of your function:
> 
> ```python
> def homographywarper(image, homo_matrix, target_h=256, target_w=256, inverse=False):
>     image_tensor = kornia.image_to_tensor(image, False).float()/255.
>     homo_matrix_tensor = torch.Tensor(homo_matrix)
>     homo_matrix_tensor = torch.unsqueeze(homo_matrix_tensor, dim=0)
>     if inverse:
>         homo_matrix_tensor = torch.inverse(homo_matrix_tensor)
>     warper = HomographyWarper(target_h, target_w, mode="nearest", align_corners=True)
>     image_tensor_warped = warper(image_tensor, homo_matrix_tensor)
>     print(image_tensor_warped.min(), image_tensor_warped.max())
>     image_warped = (tensor2image(image_tensor_warped) * 255).astype(np.uint8)
>     return image_warped
> ```
> 
> ![image](https://user-images.githubusercontent.com/4803565/122968306-b07f9c00-d38b-11eb-9834-e93b1f892743.png)

My bad! I did not realise there is a call to the third function. I remember I did use the inherent functions from Kornia, but I adjusted slightly to suit some purposes.

Anyway, thanks for figuring it out. 

Cheers, Chiang.
